File: personalizatio-sdk/src/main/java/com/personalizatio/stories/StoriesView.java
package com.personalizatio.stories;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.personalizatio.Api;
import com.personalizatio.R;
import com.personalizatio.SDK;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class StoriesView extends ConstraintLayout implements StoriesAdapter.ClickListener {

	private StoriesAdapter adapter;
	private final ArrayList<Story> list = new ArrayList<>();
	private StoryView story_view;
	private String code;

	public StoriesView(Context context, String code) {
		super(context);
		this.code = code;
	}

	public StoriesView(Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
		parseAttrs(attrs);
	}

	public StoriesView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		parseAttrs(attrs);
	}

	@TargetApi(Build.VERSION_CODES.LOLLIPOP)
	public StoriesView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
		super(context, attrs, defStyleAttr, defStyleRes);
		parseAttrs(attrs);
	}

	@Override
	protected void onFinishInflate() {
		super.onFinishInflate();
		initialize();
	}

	private void parseAttrs(AttributeSet attrs) {
		TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.StoriesView);
		code = typedArray.getString(R.styleable.StoriesView_code);
	}

	//Инициализация
	private void initialize() {
		View view = inflate(getContext(), R.layout.stories, this);
		RecyclerView stories = view.findViewById(R.id.stories);

		adapter = new StoriesAdapter(list, this);
		stories.setAdapter(adapter);

		Handler handler = new Handler(Looper.getMainLooper()) {
			@Override
			public void handleMessage(Message msg) {
				adapter.notifyDataSetChanged();
			}
		};

		//Запрашиваем сторисы
		SDK.stories(code, new Api.OnApiCallbackListener() {
			@Override
			public void onSuccess(JSONObject response) {
				Log.d("stories", response.toString());
				try {
					JSONArray json_stories = response.getJSONArray("stories");
					for( int i = 0; i < json_stories.length(); i++ ) {
						list.add(new Story(json_stories.getJSONObject(i)));
					}
					handler.sendEmptyMessage(1);
				} catch(JSONException e) {
					Log.e(SDK.TAG, e.getMessage(), e);
				}
			}
		});
	}

	@Override
	public void onStoryClick(int index) {
		Story story = list.get(index);

		//Сбрасываем позицию
		if( story.start_position >= story.slides.size() || story.start_position < 0 ) {
			story.start_position = 0;
		}

		StoryView dialog = new StoryView(getContext(), code, story, () -> {
			adapter.notifyDataSetChanged();
			if( index + 1 < list.size() ) {
				onStoryClick(index + 1);
			}
		}, () -> {
			if( index - 1 >= 0 ) {
				list.get(index - 1).start_position = list.get(index - 1).slides.size() - 1;
				onStoryClick(index - 1);
			}
		});
		dialog.show();
	}
}

File: personalizatio-sdk/src/main/java/com/personalizatio/stories/StoriesView.kt
package com.personalizatio.stories

import android.annotation.TargetApi

class StoriesView : ConstraintLayout, StoriesAdapter.ClickListener {
    private var adapter: StoriesAdapter? = null
    private val list: ArrayList<Story?>? = ArrayList()
    private var observer: ContentObserver? = null
    val settings: Settings? = Settings()
    var code: String? = null
    var player: Player? = null

    @Nullable
    var click_listener: OnLinkClickListener? = null
    var mute: Boolean = true
    var mute_listener: Runnable? = null

    constructor(context: Context?, code: String?) : super(context) {
        this.code = code
        initialize()
    }

    constructor(context: Context?, @Nullable attrs: AttributeSet?) : super(context, attrs) {
        parseAttrs(attrs)
    }

    constructor(context: Context?, @Nullable attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        parseAttrs(attrs)
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(
        context,
        attrs,
        defStyleAttr,
        defStyleRes
    ) {
        parseAttrs(attrs)
    }

    /**
     * Вызывать, когда объект сторисов удален с экрана и больше не нужен
     */
    fun release() {
        player.release()
    }

    @Override
    protected fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        unregisterObserver()
    }

    @Override
    protected fun onFinishInflate() {
        super.onFinishInflate()
        initialize()
    }

    private fun parseAttrs(attrs: AttributeSet?) {
        val typedArray: TypedArray = getContext().obtainStyledAttributes(attrs, R.styleable.StoriesView)
        code = typedArray.getString(R.styleable.StoriesView_code)
    }

    //Инициализация
    private fun initialize() {
        val view: View = inflate(getContext(), R.layout.stories, this)
        val stories: RecyclerView = view.findViewById(R.id.stories)

        adapter = StoriesAdapter(this, list, this)
        stories.setAdapter(adapter)

        val handler: Handler = object : Handler(Looper.getMainLooper()) {
            @Override
            fun handleMessage(msg: Message?) {
                registerObserver()
                adapter.notifyDataSetChanged()
            }
        }

        //Плеер для просмотра видео
        player = Player(getContext())

        settings.failed_load_text = getResources().getString(R.string.failed_load_text)

        //Запрашиваем сторисы
        SDK.stories(code, object : OnApiCallbackListener() {
            @Override
            fun onSuccess(response: JSONObject?) {
                Log.d("stories", response.toString())
                try {
                    val json_stories: JSONArray = response.getJSONArray("stories")
                    for (i in 0 until json_stories.length()) {
                        list.add(Story(json_stories.getJSONObject(i)))
                    }
                    handler.sendEmptyMessage(1)
                } catch (e: JSONException) {
                    Log.e(SDK.TAG, e.getMessage(), e)
                }
            }
        })
    }

    @Override
    fun onStoryClick(index: Int) {
        val story: Story = list.get(index)

        //Сбрасываем позицию
        if (story.start_position >= story.slides.size() || story.start_position < 0) {
            story.start_position = 0
        }

        val dialog: StoryDialog = StoryDialog(this, list, index) {
            adapter.notifyDataSetChanged()
        }
        dialog.show()
    }

    /**
     * Устанавливает слушатель клика по элементам
     * @param listener OnLinkClickListener
     */
    fun setOnLinkClickListener(@Nullable listener: OnLinkClickListener?) {
        this.click_listener = listener
    }

    fun muteVideo(mute: Boolean) {
        this.mute = mute
    }

    fun isMute(): Boolean {
        return mute
    }

    fun registerObserver() {
        observer = object : ContentObserver(Handler()) {
            @Override
            fun onChange(selfChange: Boolean) {
                super.onChange(selfChange)
                val manager: AudioManager = getContext().getSystemService(Context.AUDIO_SERVICE) as AudioManager
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    mute = manager.isStreamMute(AudioManager.STREAM_MUSIC)
                    if (mute_listener != null) {
                        mute_listener.run()
                    }
                }
            }
        }
        getContext().getContentResolver()
            .registerContentObserver(android.provider.Settings.System.CONTENT_URI, true, observer)
    }

    fun unregisterObserver() {
        if (observer != null) {
            getContext().getContentResolver().unregisterContentObserver(observer)
        }
    }
}
